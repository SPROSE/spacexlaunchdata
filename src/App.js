/* Modules */
import React, { useState, useEffect } from 'react';

/* Assets */
import logo from './assets/spacex-logo.png';
import refresh from './assets/icon/refresh.png';
import sortImg from './assets/icon/sort.png';
import spacexImage from './assets/img/launch-home.png';

/* Styles */
import './App.scss';

/* Components */
import LaunchList from './components/launch-list';
import Year from './components/year';

function App() {
    const [hasError, setErrors] = useState(false);
    const [spacexData, setSpacexData] = useState([]);
    const [yearsData, setYearsData] = useState({});
    const [sort, setSort] = useState('ascending');

    async function fetchData() {
        const res = await fetch("https://api.spacexdata.com/v3/launches");
        res
            .json()
            .then(res => reduceAndSetToState(res))
            .catch(err => setErrors(err));
    }

    function reduceAndSetToState(res) {
        const spacexDataJson = reduceJson(res);
        setSpacexData(spacexDataJson);
        const availableYears = getYears(res);
        setYearsHandler(0, availableYears, false);
    }

    const reduceJson = (res) => {
        return res.reduce((x, y) => {
            return x.concat({
                flight_number: y.flight_number,
                mission_name: y.mission_name,
                date: y.launch_date_utc,
                rocket_name: y.rocket.rocket_name,
                launch_year: y.launch_year
            });
        }, []);
    };

    const getYears = (res) => {
        const years = res.reduce((x, y) => {
            return x.concat(y.launch_year);
        }, []);

        let seen = {};
        return years.filter((item) => {
            return seen.hasOwnProperty(item) ? false : (seen[item] = true);
        });
    };

    const setYearsHandler = (year, availableYears, showYearsDropdown) => {
        const yearsData = {
            year: year,
            availableYears: availableYears,
            showYearsDropdown: showYearsDropdown
        };
        setYearsData(yearsData);
    };

    const changeSort = () => {
        setSort(sort === 'ascending' ? 'descending' : 'ascending');
    };

    const reload = () => {
        fetchData();
    };

    useEffect(() => {
        fetchData();
    }, []);

    return (
        <div className="app">
            <header className="app-header">
                <div className={"container"}>
                    <div className={"row"}>
                        <div className={"col col-7"}>
                            <img src={logo} className="logo inline bot" alt="logo" />
                            <h1 className={"page-title uppercase inline bot"}>Launches</h1>
                        </div>
                        <div className={"col col-5"}>
                            <button
                                className={"btn pull-right reload-btn"}
                                onClick={reload}>
                                Reload Data <img src={refresh} className={"icon"} alt={"refresh data"} />
                            </button>
                        </div>
                    </div>
                </div>
            </header>
            <div className={"container"}>
                <div className={"row"}>
                    <div className={"col col-5 left-image-rocket"}>
                        <img src={spacexImage} className={"spacex-image"} alt={"space x shuttle"} />
                    </div>
                    <section className={"col col-6 launch-data"}>
                        {hasError ?
                            <div>Error loading data, {hasError}</div>
                            :
                            <React.Fragment>
                                <div className={"list-controls align-right"}>
                                    <Year
                                        yearsData={yearsData}
                                        setYearsHandler={setYearsHandler}
                                    />
                                    <button
                                        className={"btn ml-1"}
                                        onClick={changeSort}>
                                        Sort {sort === 'ascending' ? 'Descending' : 'Ascending'} <img src={sortImg} className={"icon"} alt={"sort data"} />
                                    </button>
                                </div>
                                {spacexData.length ?
                                    <LaunchList
                                        year={yearsData.year}
                                        spacexData={spacexData}
                                        sort={sort}
                                        hasError={hasError}
                                    />
                                    : <div>Loading...</div> }
                            </React.Fragment>
                        }
                    </section>
                </div>
            </div>
        </div>
    );
}

export default App;
