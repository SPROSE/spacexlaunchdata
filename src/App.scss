@import './global/base/index.scss';

/* Header Custom Styles */
.app-header {
  padding: 20px 0 60px;
}

.logo {
  pointer-events: visible;
  cursor: pointer;
  max-width: 150px;
}

.reload-btn {
  padding-left: 10px;
  border-radius: 25px 0 0 25px;
}

.spacex-image {
  max-width: 100%;
}

.left-image-rocket {
  @include md {
    display: none;
  }
  @include sm {
    display: none;
  }
  @include sm-only {
    display: none;
  }
}

.launch-data {
  @include md {
    width: 100%!important;
  }
  @include sm {
    width: 100%!important;
  }
  @include sm-only {
    width: 100%!important;
  }
}


