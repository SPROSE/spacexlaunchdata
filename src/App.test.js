import React from 'react';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';

import App from './App';

const initialProps = {
    yearsData: {}
};

test('renders header elements', () => {
  const { getByText } = render(<App />);
  const headerText = getByText('Launches');
  const reloadText = getByText('Reload Data');
  expect(headerText).toBeInTheDocument();
  expect(reloadText).toBeInTheDocument();
});

describe('<App /> with props', () => {
    const container = shallow(<App {...initialProps} />);
    it('should match the snapshot', () => {
        expect(container.html()).toMatchSnapshot();
    });
});