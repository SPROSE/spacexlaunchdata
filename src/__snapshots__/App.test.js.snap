// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> with props should match the snapshot 1`] = `"<div class=\\"app\\"><header class=\\"app-header\\"><div class=\\"container\\"><div class=\\"row\\"><div class=\\"col col-7\\"><img src=\\"spacex-logo.png\\" class=\\"logo inline bot\\" alt=\\"logo\\"/><h1 class=\\"page-title uppercase inline bot\\">Launches</h1></div><div class=\\"col col-5\\"><button class=\\"btn pull-right reload-btn\\">Reload Data <img src=\\"refresh.png\\" class=\\"icon\\" alt=\\"refresh data\\"/></button></div></div></div></header><div class=\\"container\\"><div class=\\"row\\"><div class=\\"col col-5 left-image-rocket\\"><img src=\\"launch-home.png\\" class=\\"spacex-image\\" alt=\\"space x shuttle\\"/></div><section class=\\"col col-6 launch-data\\"><div class=\\"list-controls align-right\\"><div class=\\"relative inline\\"><button class=\\"btn\\">Filter by Year <img src=\\"select.png\\" class=\\"icon\\" alt=\\"select year\\"/></button></div><button class=\\"btn ml-1\\">Sort Descending <img src=\\"sort.png\\" class=\\"icon\\" alt=\\"sort data\\"/></button></div><div>Loading...</div></section></div></div></div>"`;
